Add mode to TarArchiveReader

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63332 Add mode to TarArchiveReader**

Add a corresponding PR from [torchdata](https://github.com/facebookexternal/torchdata/pull/101)

Differential Revision: [D30350151](https://our.internmc.facebook.com/intern/diff/D30350151)

ejguan committed:
Add mode to TarArchiveReader

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
please sync with torchdata

ejguan commented:
> please sync with torchdata

It's a sync PR from torchdata.

ejguan committed:
Update on "Add mode to TarArchiveReader"


Add a corresponding PR from [torchdata](https://github.com/facebookexternal/torchdata/pull/101)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30350151).
